ITSI Login Error

rajadain opened the pull request:
This PR presents a tentative approach for route based error handling, with the immediate application of handling route based errors due to ITSI OAuth failures. It can be outlined as follows:

 * All `/error[.*]` routes will be handled by Django to render the main Backbone app
 * The Backbone app will pass these routes to an `ErrorController`
 * The `ErrorController` has a list of error `types` from which it looks up a specific `ErrorHandler` and fires it.

The following questions are raised by this work:

 * Currently the `ErrorHandler` for ITSI just pops up an `alert` message, which is rather inelegant. We should probably have design guidelines for showing error messages. Created #224 for this.
 * What should happen if the `ErrorController` is not passed in any error `type`, e.g. `/error` or `/error/`? Currently this is handled by `ErrorHandler.generic` and pops up a bland `alert` box.
 * What should happen if the `ErrorController` cannot find a matching error `type`, e.g. `/error/xyz`? Currently this is handled by `ErrorHandler.unknown` and silently logs the error to the console.

If we find `ErrorController` growing into having models and views, we might consider moving it out of the generic `controllers.js` into its own folder.

This PR addresses part of #180, namely the error view. I will wait for work on #152 to finish before adding a registration view.

Connects #180

mmcfarland commented on src/mmw/js/src/controllers.js:
```diff
@@ -26,6 +27,34 @@ var AppController = {
+var ErrorHandlers = {
+    itsi: function() {
+        router.navigate('');
+        alert("We're sorry, but there was an error logging you in with your" +
```
As part of this PR, can you create an issue for @ctaylo37 to give us some markup for a general purpose error modal?

rajadain replied:
Just added #224.

caseypt commented:
When you get in this morning, can you walk me through this PR? I'm a little out-of-date on the ITSI stuff.

caseypt commented on src/mmw/js/src/controllers.js:
```diff
@@ -26,6 +27,44 @@ var AppController = {
 module.exports = {
-    AppController: AppController
+    AppController: AppController,
+    ErrorController: ErrorController
```
I was hoping to phase out this top-level controller file (we've been moving the page routes to their respective apps as they've been created). Maybe we could move this to `core/errorController.js`? Curious what others think.

caseypt replied:
Just saw your comment about moving this to it's own folder. I would consider `errorViews`, `errorModels`, `errorController`, etc in `core` before moving it to it's own app. But again, open to input.

rajadain replied:
I agree, cross-cutting concerns like error handling and logging should be in core, not in separate apps. I do like the simplicity of one-word non-CamelCase file names we have so far though, so perhaps we could do `core/error/views.js` and `/core/error/controllers.js` instead?

caseypt replied:
I'm cool with this. Any thoughts @mmcfarland?

mmcfarland replied:
Yeah, agree that `core` is the right place, the sub-dir `error` seems like a good layout to me.

caseypt commented:
Here's what I'm seeing. Is this the correct behavior?

- Go to localhost:8000
- Navigate to user/itsi/login
- Login on ITSI using the `tstudent2` credentials.
- Get redirected back to localhost:8000
- An alert pops up with the following message: "We're sorry, but there was an error logging you in with your ITSI credentials. Please log in using another method or continue as a guest."

rajadain commented:
Well, you were redirected to `/error/itsi` and it showed you the popup message, so that part of the PR works, which is good. But no, since authentication was successful you should've been re-routed to `/register` instead.

Can you try running migrations to make sure that the `user_itsiuser` table is setup?

rajadain commented:
Also please make sure you have setup the `MMW_ITSI_SECRET_KEY` environment variable in the `app` VM, as described in the README. That could cause errors when negotiating an OAuth token.

caseypt commented:
Thanks. The env var fixed it. I was successfully redirected to `/register`. 

Without the env var, I inadvertently tested that the error handling does work.

+1

rajadain commented:
As a side note, to test successful user authentication with ITSI account `tstudent2`, please do the following in the app shell `scripty manage shell`:

```python
from django.contrib.auth.models import User
from apps.user.models import ItsiUser
user = User.objects.create_user(username="tstudent2", email=None, password=None)
itsi_user = ItsiUser.objects.create_itsi_user(user, 214)
user.save()
itsi_user.save()
```

caseypt commented:
Cool. Ran that script, logged in, and was successfully redirected to the index page instead of register.

kdeloach commented:
+1

rajadain commented:
I've pushed some new commits to this PR, based on feedback from @caseypt, @kdeloach and @mmcfarland. Changes are as follows:

 * Moved `ErrorController` and `ErrorHandlers` out of `src/controllers.js` and into `src/core/error/controllers.js` and `src/core/error/handlers.js` respectively.
 * Combined `generic` and `unknown` error handlers into one
 * Removed awkward code for capturing URL before `Backbone.history.start` since that is handled natively by Backbone

Resulting code is much sleeker, thanks for all your input. Will merge in shortly.
